src/types/chat.rs

```diff
@@ -0,0 +1,445 @@
+use std::collections::{BTreeMap, BTreeSet};
+use std::error::{Error, FromError};
+use std::io;
+
+use rustc_serialize::json;
+use rustc_serialize::json::{Json, ToJson};
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum ChatJsonError {
+    MalformedJson(json::ParserError),
+    IoError(io::Error),
+    NotAnObject,
+    InvalidFieldType,
+    UnknownField(String),
+    InvalidColor(String),
+    InvalidClickEvent,
+    InvalidHoverEvent
+}
+
+impl FromError<io::Error> for ChatJsonError {
+    fn from_error(err: io::Error) -> ChatJsonError {
+        ChatJsonError::IoError(err)
+    }
+}
+
+impl FromError<json::ParserError> for ChatJsonError {
+    fn from_error(err: json::ParserError) -> ChatJsonError {
+        if let json::ParserError::IoError(e) = err {
+            ChatJsonError::IoError(e)
+        } else {
+            ChatJsonError::MalformedJson(err)
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ChatJson {
+    pub msg: Message,
+    pub extra: Option<Vec<Json>>,
```

Comment:
Ah, that wasn't clear to me from the spec. I think I need some more test cases in order to write the desugaring logic, and maybe I could do it recursively.